### PR TITLE
Feature/freep 602

### PR DESCRIPTION
**Pull request by AlekseiMisyukevich**

### CODE REVIEW

File: contracts/Sale.sol [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    struct Nft {
+        uint256 price;
+        uint256 quantity;
+    }
+
+  	/**
+  	 * Mapping represents relation such as 
+     * seller address => (nft ID => structure containing price, quantity) 
+  	 */
+  	mapping(address => mapping(uint => Nft)) nftPrice;
```
Comment (naure) [inline contracts/Sale.sol:24]: Is this feature an upgrade for Davinci or for the new major release of Freeport?

If the former, then what happens to the existing offers?

**Review by naure — CHANGES_REQUESTED** [review]
Most changes here are not related to FREEP-602. Can you make a minimal PR with only the relevant changes per feature?

File: contracts/Sale.sol (hunk 1 above)
Comment (AlekseiMisyukevich) [reply contracts/Sale.sol:24]: will be a release if the new bid structure. past bids will be lost
though I can make it back compatible by storing quantity in another mapping. Which should I go with?

File: contracts/Sale.sol [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+     * a race condition with makeOffer or setExchangeRate. Pass 0 to disable this validation and accept any current price.
+     */
+    function takeOffer(address buyer, address seller, uint256 nftId, uint256 expectedPriceOrZero, uint256 amount)
+    public {
+        Nft storage nft = nftPrice[seller][nftId];
+        uint256 price = nft.price;
+        address payer = _msgSender();
+        require(price != 0, "Not for sale");
+        require(expectedPriceOrZero == 0 || expectedPriceOrZero == price, "Unexpected price");
+        require(nft.quantity - amount >= 0, "Remaining quantity exceeded");
```
[the remark this answers is not in the document]
Comment (AlekseiMisyukevich) [reply contracts/Sale.sol:93]: It reverts state if difference is negative. It's correct

File: contracts/Sale.sol [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+contract Sale is SaleBase {
+
+  	function initialize(Freeport freeport, address stablecoin) public initializer {
+  	  __SaleBase_init(freeport, stablecoin);
+  	}
+
+    /**
+     * Stores nft id and its quantity for sale. 
+     */
+    struct Nft {
```
[the remark this answers is not in the document]
Comment (AlekseiMisyukevich) [reply contracts/Sale.sol:15]: please advise different name if this structure is member of new contract

File: contracts/Sale.sol (hunk 1 above)
Comment (AlekseiMisyukevich) [reply contracts/Sale.sol:24]: It's actually new deployment. Don't think we can link it with freeport storage

Comment (naure) [reply contracts/Sale.sol]: Ok, so this PR must go into the next major release (incompatible).

**Comment by AlekseiMisyukevich** [comment]
> Most changes here are not related to FREEP-602. Can you make a minimal PR with only the relevant changes per feature?

u right, I included too much accidentely

**Comment by naure** [comment]
There is no such thing as a too small PR ;)

**Comment by AlekseiMisyukevich** [comment]
let's close this one. I'll include freep-602 in one of freep-555 PRs that brings Sale contract, ok?

**Comment by AlekseiMisyukevich** [comment]
Adding quantity to older older SimpleExchange is meaningless because I'll have to extract and refactor SimpleExchange and related tests twice. 2nd time comes after Sale contract implementation. So it's more effective to provide quantity feature with Sale contract.

**Pull request closed without merging**